contracts/FlashRolloverV1toV2.sol

```diff
@@ -0,0 +1,364 @@
+        ADDRESSES_PROVIDER = _addressesProvider;
+        LENDING_POOL = ILendingPool(_addressesProvider.getLendingPool());
+
+        owner = msg.sender;
+    }
+
+    function rolloverLoan(
+        RolloverContractParams calldata contracts,
+        uint256 loanId,
+        LoanLibrary.LoanTerms calldata newLoanTerms,
+        address lender,
+        uint160 nonce,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external override {
+        LoanLibraryV1.LoanTerms memory loanTerms = contracts.sourceLoanCore.getLoan(loanId).terms;
+
+        {
+            _validateRollover(contracts.sourceLoanCore, contracts.targetVaultFactory, loanTerms, newLoanTerms, contracts.sourceLoanCore.getLoan(loanId).borrowerNoteId);
+        }
+
+        {
+            address[] memory assets = new address[](1);
+            assets[0] = loanTerms.payableCurrency;
+
+            uint256[] memory amounts = new uint256[](1);
+            amounts[0] = loanTerms.principal + loanTerms.interest;
+
+            uint256[] memory modes = new uint256[](1);
+            modes[0] = 0;
+
+            bytes memory params = abi.encode(
+                OperationData({ contracts: contracts, loanId: loanId, newLoanTerms: newLoanTerms, lender: lender, nonce: nonce, v: v, r: r, s: s })
+            );
+
+            // Flash loan based on principal + interest
+            LENDING_POOL.flashLoan(address(this), assets, amounts, modes, address(this), params, 0);
+        }
+    }
+
+    function executeOperation(
+        address[] calldata assets,
+        uint256[] calldata amounts,
+        uint256[] calldata premiums,
+        address initiator,
+        bytes calldata params
+    ) external override nonReentrant returns (bool) {
+        require(msg.sender == address(LENDING_POOL), "unknown callback sender");
+        require(initiator == address(this), "not initiator");
+
+        return _executeOperation(assets, amounts, premiums, abi.decode(params, (OperationData)));
+    }
+
+    function _executeOperation(
+        address[] calldata assets,
+        uint256[] calldata amounts,
+        uint256[] calldata premiums,
+        OperationData memory opData
+    ) internal returns (bool) {
+        OperationContracts memory opContracts = _getContracts(opData.contracts);
+
+        // Get loan details
+        LoanLibraryV1.LoanData memory loanData = opContracts.loanCore.getLoan(opData.loanId);
+
+        address borrower = opContracts.borrowerNote.ownerOf(loanData.borrowerNoteId);
+
+        // Do accounting to figure out amount each party needs to receive
+        (uint256 flashAmountDue, uint256 needFromBorrower, uint256 leftoverPrincipal) = _ensureFunds(
+            amounts[0],
+            premiums[0],
+            opContracts.feeController.getOriginationFee(),
+            opData.newLoanTerms.principal
+        );
+
+        IERC20 asset = IERC20(assets[0]);
+
+        if (needFromBorrower > 0) {
```

Comment:
Instead of an if statement here, could we do a require statement that needFromBorrower is gt 0? Not sure what would be the outcome if equals 0

Reply:
`needFromBorrower` doesn't need to be gt 0, if it's 0 that's OK. But if it _is_ gt 0, we should perform those extra require checks on the next two lines. Putting `needFromBorrower` into the checks themselves makes them a little verbose and redundant IMO